backend/src/apiserver/config/sample_config.json

```diff
@@ -1,4 +1,9 @@
 [
+  {
+    "name": "[Demo] XGBoost - Iterative model training",
+    "description": "[source code](https://github.com/kubeflow/pipelines/blob/master/samples/core/train_until_good/train_until_good.py) This sample demonstrates iterative training using a train-eval-check recursive loop. The main pipeline trains the initial model and then gradually trains the model some more until the model evaluation metrics are good enough.",
+    "file": "/samples/core/train_until_good/train_until_good.py.yaml"
```

Comment:
I heard from previous discussion, using separate steps to achieve train-eval-check loop is not suggested for perf reasons. Does it not apply for XGBoost?

Reply:
Let's think about general idea of pipelines and the value they bring to the users.
People can already do data transformation, training, prediction and evaluation without Pipelines. They can just copy pieces of code into a notebook and run that end-to-end code. So why would they use KFP? There are several advantages. KFP promotes componentization and granularity. Different components can be composed together to build a higher-level pipeline. Having granularity is also useful due to caching of outputs of individual steps.

We want to show users how to detangle monolithic training scripts into flexible pipelines. This is why we demonstrate separation and composition.

Doing train-eval-check loop inside a single task is possible, but that does not demonstrate KFP in the best way. This also limits what the user can do. In that case, when the user needs something that's not already provided, they are blocked. For example, there are many metrics not included in XGBoost. There are use cases where the user needs to calculate metrics over multiple datasets (for example, new model on new dataset vs prod model on new dataset vs prod model on old dataset). There are also metrics that XGBoost cannot possibly calculate. For example, GCP, AWS and Azure have labelling services. To calculate real metrics for some of ML tasks (for example, ranking), new labels need to be acquired from humans or usage statistics. It's hard and prohibitively expensive to calculate such metrics during the training. This is why this pipeline demonstrates the template for a batched training.

Think of most of our sample pipelines. For example the coin toss pipeline. You do not need Kubernetes or `dsl.Conditional` for such a pipeline. But the coin toss pipeline demonstrates important KFP constructs. The same is true about this demo pipeline. It shows the users what's possible with KFP and helps them build their own pipelines.

Reply:
Thanks Alexey, these are great points!